diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index df2b2bb9fa5d81..6e9ca88aa498d6 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -2533,6 +2533,9 @@ Kernel parameters
 			disable
 			  Do not enable intel_pstate as the default
 			  scaling driver for the supported processors
+			enable
+			  Enable intel_pstate in-case "disable" was passed
+			  previously in the kernel boot parameters
                         active
                           Use intel_pstate driver to bypass the scaling
                           governors layer of cpufreq and provides it own
diff --git a/drivers/cpufreq/intel_pstate.c b/drivers/cpufreq/intel_pstate.c
index 11c58af4190064..6eb8ec3f278d87 100644
--- a/drivers/cpufreq/intel_pstate.c
+++ b/drivers/cpufreq/intel_pstate.c
@@ -3916,6 +3916,8 @@ static int __init intel_pstate_setup(char *str)
 
 	if (!strcmp(str, "disable"))
 		no_load = 1;
+	else if (!strcmp(str, "enable"))
+		no_load = 0;
 	else if (!strcmp(str, "active"))
 		default_driver = &intel_pstate;
 	else if (!strcmp(str, "passive"))
diff --git a/drivers/acpi/processor_idle.c b/drivers/acpi/processor_idle.c
index f6c72e3a2be1b5..c2afe55be63133 100644
--- a/drivers/acpi/processor_idle.c
+++ b/drivers/acpi/processor_idle.c
@@ -1409,6 +1409,17 @@ void acpi_processor_power_init(struct acpi_processor *pr)
 	if (disabled_by_idle_boot_param())
 		return;
 
+	/*
+	 * The bus master check on AMD is observed to be very costly and results
+	 * in not only increased power consumption, but also lots of CPU cycles
+	 * burned. Disable the actual bus master check while keeping
+	 * pr->flags.bm_check enabled so there's no cache flush upon C3 entry.
+	 */
+	if (!bm_check_disable && boot_cpu_data.x86_vendor == X86_VENDOR_AMD) {
+		bm_check_disable = true;
+		pr_info("Enabled processor.bm_check_disable for AMD cpuidle\n");
+	}
+
 	acpi_processor_cstate_first_run_checks();
 
 	if (!acpi_processor_get_power_info(pr))
diff --git a/drivers/acpi/processor_idle.c b/drivers/acpi/processor_idle.c
index c2afe55be63133..74f91ed3db1625 100644
--- a/drivers/acpi/processor_idle.c
+++ b/drivers/acpi/processor_idle.c
@@ -1409,6 +1409,7 @@ void acpi_processor_power_init(struct acpi_processor *pr)
 	if (disabled_by_idle_boot_param())
 		return;
 
+#ifdef CONFIG_X86
 	/*
 	 * The bus master check on AMD is observed to be very costly and results
 	 * in not only increased power consumption, but also lots of CPU cycles
@@ -1419,6 +1420,7 @@ void acpi_processor_power_init(struct acpi_processor *pr)
 		bm_check_disable = true;
 		pr_info("Enabled processor.bm_check_disable for AMD cpuidle\n");
 	}
+#endif
 
 	acpi_processor_cstate_first_run_checks();
 
